### PR TITLE
copier code refine and module parameter simplify

### PR DESCRIPTION
**Pull request by btian1**
@kv2019i note: this is split from https://github.com/thesofproject/sof/pull/7909

### CODE REVIEW

File: src/audio/eq_fir/eq_fir.c
Comment (lyakh) [inline src/audio/eq_fir/eq_fir.c:180]: before this change `mod->stream_params` wasn't modified, it was used as a template, now it's modified. Is this correct?

File: src/audio/eq_iir/eq_iir.c
Comment (lyakh) [inline src/audio/eq_iir/eq_iir.c:844]: same here

File: src/audio/src/src.c
Comment (lyakh) [inline src/audio/src/src.c:541]: `mod->stream_params` is modified

**Review by kv2019i — COMMENTED** [review]
Thanks for splitting this out @btian1 ! I think this is very good cleanup for the most part. There are a few smaller things were we change functionality and it's not a plain refactoring. I'm ok with this but please @ranj063 ack these are correct.

And a generic rant and reminder that please let's remember to add rationale to git commit message if we a change is not a simpler refactor.

File: src/audio/module_adapter/module/generic.c
Comment (ranj063) [reply src/audio/module_adapter/module/generic.c:91]: @kv2019i this is safe to remove for IPC3 but as for IPC4, I think i'd like @dbaluta to confirm is it is OK to remove for the cadence codecs

File: src/audio/copier/copier_dai.c
Comment (RanderWang) [inline src/audio/copier/copier_dai.c:268]: please drop this commit. It will be used by https://github.com/thesofproject/sof/pull/7959/commits/78a8eec9cb9a010065c3bd7294f699986fbe7015

**Review by RanderWang — COMMENTED** [review]
Good job!  Please check my comments about create_endpoint_buffer

File: src/audio/eq_fir/eq_fir.c
Comment (lyakh) [reply src/audio/eq_fir/eq_fir.c:180]: I'm confused so I'm going to give this PR a -1 for now, I think we have to clarify something. I tried to follow where parameters come from and how they're used, so I went back to https://github.com/thesofproject/sof/commit/a25cdbb90b67be1f08c81e8de1641b7cefdf5fd5 and I now don't think it's quite correct. At the very least it doesn't seem to do what its description is saying. It isn't updating parameters, it's degrading them. Parameters are allocated in `module_adapter_params()` and copied from the IPC, but then that commit overwrites it with original data from `.base_cfg`. Now, I think with IPC4 data "from the IPC" is actually fake and all 0 - it comes from https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc4/handler.c#L117 but then some of it is overwritten in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/audio/module_adapter/module_adapter.c#L522 and also some of it is used to configure buffers e.g. in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc-helper.c#L140 or in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc-helper.c#L156 . Then parameters are overwritten, but buffer parameters remain as they were set in `comp_verify_params()`... At least should make it obvious in `module_adapter_params()` and only use that data for IPC3? Then, those two added lines inhttps://github.com/thesofproject/sof/commit/a25cdbb90b67be1f08c81e8de1641b7cefdf5fd5 - they fill the same data into two locations - can we just use `memcpy()` for the second one? And do we actually need to write data to `params` - it's a local variable in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc4/handler.c#L117 and it will be used for all components in the pipeline while recursing in https://github.com/thesofproject/sof/blob/214f98378f12ed179bd22f0c4feb09b0b3aaea81/src/ipc/ipc4/handler.c#L130 so every component (if it's using the module adapter API) will just overwrite it with its own copy...
Sorry for a long and not very consistent comment, but I think we need to clean up and clarify a bit before we add more complexity to parameter update and propagation flow

**Review by lyakh — CHANGES_REQUESTED** [review]
Can we hold on with this a bit until we clarify my doubts about parameter setting. I'll give this -1 for now just to make sure that we don't forget to answer these questions before merging

File: src/audio/aria/aria.c
Comment (singalsu) [inline src/audio/aria/aria.c:193]: Nice, I didn't know about this function! It helps to simplify other module adapter conversion patches.

File: src/audio/eq_fir/eq_fir.c
Comment (btian1) [reply src/audio/eq_fir/eq_fir.c:180]: @lyakh , @kv2019i  really thanks you bring so much for this topic, how to say, like I may have similar feelings like you, just can't change too much a time(CI need pass), I have below things want to clarify:

1. Nobody said the logic can't change, yes, params are changing compared with previous.
2. there are also other functions to change params, like comp_verify_params, comp_update_params.

so, I will change first when there is a reason and check CI and then your comments, somehow, CI is first gold criteria for me.

I can remove these part of changes first, and then discuss further for the actual changes to unblock the PR merge, let me know what do you think.

File: src/audio/module_adapter/module/generic.c
Comment (btian1) [reply src/audio/module_adapter/module/generic.c:91]: @dbaluta , will IPC3 also support cadence codecs?

File: src/audio/eq_fir/eq_fir.c
Comment (lyakh) [reply src/audio/eq_fir/eq_fir.c:180]: @btian1 in fact I'd even try to make a PR to clean up and clarify my above comments and to add any of your fixes to that parameter handling code, if needed. And then update this PR on top of that one. I think that order would be better because fixes should come before optimisations. Alternatively you could still remove those changes from this PR and add a PR after this one, but that would probably (1) first add more confusion to the current situation in this PR, and (2) make it then more difficult to fix it in a follow-up.

Comment (btian1) [reply src/audio/eq_fir/eq_fir.c:180]: ok, let me remove params related patches as first step to make this PR can be reviewed and merged, and we can discuss later for params related changes.

**Review by lyakh — APPROVED** [review]

File: src/audio/module_adapter/module/generic.c
Comment (dbaluta) [reply src/audio/module_adapter/module/generic.c:91]: @btian1  yes!

**Comment by dbaluta** [comment]
@btian1 will give it a test today!

File: src/audio/module_adapter/module/generic.c
Comment (btian1) [reply src/audio/module_adapter/module/generic.c:91]: so we can't remove those ipc3 specific code? @dbaluta @lyakh @ranj063, since ipc3 need it? from my observation, I did not see those code is must even for IPC3

**Review by dbaluta — CHANGES_REQUESTED** [review]
This PR doesn't compile with IPC3

```
/home/nxf25322/work/sof_dir/sof/src/audio/module_adapter/module/cadence.c: In function 'cadence_codec_init':
/home/nxf25322/work/sof_dir/sof/src/audio/module_adapter/module/cadence.c:356:11: error: 'struct module_data' has no member named 'state'
  if (codec->state == MODULE_DISABLED && codec->cfg.avail) {
           ^~
/home/nxf25322/work/sof_dir/sof/src/audio/module_adapter/module/cadence.c:356:22: error: 'MODULE_DISABLED' undeclared (first use in this function); did you mean 'LOG_DISABLE'?
  if (codec->state == MODULE_DISABLED && codec->cfg.avail) {
                      ^~~~~~~~~~~~~~~
                      LOG_DISABLE
/home/nxf25322/work/sof_dir/sof/src/audio/module_adapter/module/cadence.c:356:22: note: each undeclared identifier is reported only once for each function it appears in
/home/nxf25322/work/sof_dir/sof/src/audio/module_adapter/module/cadence.c: In function 'cadence_codec_set_configuration':
/home/nxf25322/work/sof_dir/sof/src/audio/module_adapter/module/cadence.c:855:22: error: unused variable 'md' [-Werror=unused-variable]
  struct module_data *md = &mod->priv;
                      ^~
cc1: all warnings being treated as errors
make[2]: *** [CMakeFiles/sof.dir/build.make:1123: CMakeFiles/sof.dir/src/audio/module_adapter/module/cadence.c.o] Error 1

```

File: src/audio/google/google_rtc_audio_processing.c
Comment (dbaluta) [inline src/audio/google/google_rtc_audio_processing.c:113]: Here you assume that google rtc audio always uses IPC4 right?

File: src/audio/module_adapter/module/generic.c
Comment (dbaluta) [reply src/audio/module_adapter/module/generic.c:91]: @btian1 we can't remove IPC3 specific code as IMX uses it. 

This patch works fine for imx cadence tests. You just need to do the following changes:

```
diff --git a/src/audio/module_adapter/module/cadence.c b/src/audio/module_adapter/module/cadence.c
index ef912d7e2..dd0654228 100644
--- a/src/audio/module_adapter/module/cadence.c
+++ b/src/audio/module_adapter/module/cadence.c
@@ -353,7 +353,7 @@ static int cadence_codec_init(struct processing_module *mod)
        codec->mpd.init_done = 0;
 
        /* copy the setup config only for the first init */
-       if (codec->state == MODULE_DISABLED && codec->cfg.avail) {
+       if (codec->cfg.avail) {
                setup_cfg = &cd->setup_cfg;
 
                /* allocate memory for set up config */
@@ -852,7 +852,6 @@ cadence_codec_set_configuration(struct processing_module *mod, uint32_t config_i
                                const uint8_t *fragment, size_t fragment_size, uint8_t *response,
                                size_t response_size)
 {
-       struct module_data *md = &mod->priv;
        struct comp_dev *dev = mod->dev;
        int ret;
```

**Review by lgirdwood — COMMENTED** [review]
@btian1 it may be simpler here for any IPC changes to split between files i.e.
we could have file-generic.c (core and generic feature code) with file-ipc3.c and file-ipc4.c (for IPC specifc feature adaptions).

File: src/audio/module_adapter/module/generic.c
Comment (btian1) [reply src/audio/module_adapter/module/generic.c:91]: @dbaluta yes, I did it last night, if this patch works fine for imx ipc3 cadence codec test, means md->state can be removed for both ipc4 and ipc3, if so, could you add your approval to this PR?

@lgirdwood ack, due to CI does not check ipc3 build, some place change are ignored, I added last night.

This change is not a irrelevant change, it is critical for host performance in #7909

**Comment by btian1** [comment]
> 

ok, if can't remove it, we have to split it to two files with ipc3 / ipc4 prefix file.

**Comment by btian1** [comment]
I will split one patch with ipc3 and ipc4, please stay tuned for next round review.

**Comment by lyakh** [comment]
> ok, if can't remove it, we have to split it to two files with ipc3 / ipc4 prefix file.

@btian1 sorry, could you explain? Do I understand it correctly, that you want to split a file into two parts because you cannot remove the `mod->priv.state` field? Isn't that already mostly encapsulated in `#if CONFIG_...IPC...3` so you can just add a couple of missing ones and the effect of handling that `.state` field should anyway be negligible?

**Comment by lgirdwood** [comment]
> > ok, if can't remove it, we have to split it to two files with ipc3 / ipc4 prefix file.
> 
> @btian1 sorry, could you explain? Do I understand it correctly, that you want to split a file into two parts because you cannot remove the `mod->priv.state` field? Isn't that already mostly encapsulated in `#if CONFIG_...IPC...3` so you can just add a couple of missing ones and the effect of handling that `.state` field should anyway be negligible?

You need to do whatever you think will be best here for simplicity, readability, velocity - it could be separate files for ipc OR could be using ifdefs. The main point is you cant break IPC3 and the code must make sense, be maintainable, readable etc.

Reviewers can then help align and fine tune the PR so it's acceptable to everyone.

**Comment by btian1** [comment]
@lgirdwood @dbaluta @ranj063  I removed the md->state patch, and I plan submit another pr to cover audio/src/module_adapter/module related ipc3/ipc4 split, please review again, this pr should be ready for merge.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@dbaluta good for you now ?

**Review by juimonen — APPROVED** [review]
changes look fine with me if CI is happy...

File: src/audio/copier/copier_dai.c
Comment (ranj063) [inline src/audio/copier/copier_dai.c:34]: a part of this is in the previous commit. Can you please reorganize to move all of the same change to the same patch?

File: src/audio/copier/copier_dai.c
Comment (ranj063) [inline src/audio/copier/copier_dai.c]: this needs to be a separate patch with an explanation for why it is not needed anymore

File: src/audio/copier/copier_dai.c
Comment (ranj063) [inline src/audio/copier/copier_dai.c]: @btian it will be better if you split this PR into 2 , one with the copier clean up and the other for using the update_buffer_format function for the other modules

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:104]: done

File: src/audio/copier/copier_dai.c
Comment (btian1) [reply src/audio/copier/copier_dai.c:34]: done

File: src/audio/copier/copier_dai.c
Comment (btian1) [reply src/audio/copier/copier_dai.c]: done

File: src/audio/copier/copier_dai.c
Comment (btian1) [reply src/audio/copier/copier_dai.c]: outdated, all changes should be done now.

**Review by ranj063 — APPROVED** [review]

**Review by iuliana-prodan — APPROVED** [review]

**Comment by lgirdwood** [comment]
@dbaluta good for you - assuming so as @iuliana-prodan has approved ?

**Review by cujomalainey — APPROVED** [review]

**Comment by btian1** [comment]
> @dbaluta good for you - assuming so as @iuliana-prodan has approved ?

yes, Prodan helped verify and approved, since Baluta is taking annual leave.

**Comment by kv2019i** [comment]
https://sof-ci.01.org/sofpr/PR7958/build11365/devicetest/index.html and https://sof-ci.01.org/sofpr/PR7958/build11366/devicetest/index.html are ok, with known fails only. With @iuliana-prodan 's ack in place of @dbaluta , I'll proceed with merge.

**Pull request merged**
